### PR TITLE
storage: Cleanup cacher testing

### PR DESCRIPTION
**Pull request by MadhavJivrajani**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR cleans up the testing dir of the cacher and enables cacher tests to reside with the cacher code.

#### Commit 1:

This commit prepares for when cacher tests are moved here
from the `tests` package. Tests in that package redeclare
some of the testing utils that exist here, so dedup-ing them.

This commit also adapts to any changes in test util signatures.
There are still some utils that can be reused but currently are
highly specific to some tests. (ex: watch_cache_test.go)

#### Commit 2

Doing this allows us to implement some more nuanced cacher manipulations
to be used in testing. For ex: implementing a test-only compaction method
for the watch cache.

#### Which issue(s) this PR fixes:

Towards https://github.com/kubernetes/kubernetes/pull/116178#discussion_r1219088283
Towards https://github.com/kubernetes/kubernetes/issues/109831
From https://github.com/kubernetes/kubernetes/pull/116178#discussion_r1219088283

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/assign @wojtek-t

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_testing_utils_test.go
Comment (MadhavJivrajani) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_testing_utils_test.go:1]: This is where I'm hoping we can put the cacher implementation with the compaction method.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 1]
```diff
@@ -762,27 +746,6 @@ func TestCacherNoLeakWithMultipleWatchers(t *testing.T) {
 	}
 }
 
-func TestWatchInitializationSignal(t *testing.T) {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:765]: Why did you remove this test?

Comment (MadhavJivrajani) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:765]: @wojtek-t - mainly because this is already a part of `cacher_test.go`: https://github.com/kubernetes/kubernetes/blob/4c40d749006a8895f6718f2b624a3dbe975988ab/staging/src/k8s.io/apiserver/pkg/storage/tests/cacher_test.go#L383

From both the implementations, at least based on what it is testing for, I thought it was semantically equivalent and hence could be removed. The only difference I could see is the new test (using the common tester) is propagating an object and creating a watch against the created object's rv, and the old test is creating one against `0`. Am I missing something?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:765]: Ok - you're right. I forgot about that.

**Comment by MadhavJivrajani** [comment]
Thanks!
/priority important-longterm
